### DISCUSSION TITLE
Best practices for high-dimensional problems

### DISCUSSION THREAD

**Discussion by homerjed** [NIFTy-PPL/NIFTy#41] [Q&A]
Hi all,

I've been experimenting with deep generative priors in JIFTY. As you can see from the example, it works very nicely. This is very similar to Knollmüller and Enßlin's Deep Bayesian Reasoning paper (they went up to 512 x 512 x 3 image problems).

I simply wrap my generative model in a `jft.WrappedCall` and optimise with MGVI, works very nicely (figure shows signal, data, reconstruction).

<img width="610" alt="image" src="https://github.com/user-attachments/assets/da03b314-0ac7-4635-8e1e-f032beebbb43" />

I can only just execute this optimisation on a single GPU given the memory constraints. I'd like to add more samples during the optimisation and go to higher dimensions with non-linear resampling and maybe GeoVI.

This problem has 64 x 64 x 3 latent parameters and I can only do sample twice during the optimisation (including antithetic samples). I worry that going to higher dimensions is simply impossible because the model and its gradients within the optimisation loop require too much memory...

I guess my question is, what are some best practices here? I notice you have a `Vmap` model in the library. Also, what is your opinion regarding the `jax.sharding` api? I have a sharding wrapper for my sample function that I think shards the latent variables sampled by MGVI but I haven't tried it yet since I think the size of the model is the problem!

Cheers!

**Comment by ph-frank** [comment]
Hi,
Thanks for opening this discussion! Also really, really cool application! :) Things on this topic in `nifty.re` are still on the move (particularly regarding `jax.sharding`, see below), but let me briefly comment on it anyway:
- There is development work underway to support some forms of sharding (particularly over samples) in `jft.opmimize_kl` (see branch [`multi_gpu`](https://gitlab.mpcdf.mpg.de/ift/nifty/-/tree/multi_gpu?ref_type=heads) on the Max-Planck gitlab). Feel free to test and have a look if this work is suitable for your needs, also tagging @roth-jakob here who is working on this one.
- Some practices/experiences regarding memory: `jax.jit` inlines things that can be treated as constant in a forward pass (e.g. all parameters of a trained neural network that are not part of the optimization for the networks input layer). This can lead to multiple copies being made during a typical nifty VI run, increasing memory. To avoid this, we allow (and for large models encourage) to declare all constants that should not be inlined as non-static. To achieve this, you need to opt for the slightly more verbose features of `jft.Model` directly instead of using `jft.WrappedCall` only. Specifically this is done by creating a model class that inherits from `jft.Model` and declare all attributes as `non-static` via their metadata. As an example, have a look at the attributes of `jft.Gaussian` (such as [jft.Gaussian.data](https://github.com/NIFTy-PPL/NIFTy/blob/2fa69f8613b430287e0cf27b4893349ae74ab93d/src/re/likelihood_impl.py#L104)).

Hope this helps! Let me know if you have any more follow up questions.

**Comment by roth-jakob** [comment]
Hi,
As mentioned by @ph-frank I am currently developing the option to map the samples of `optimize_kl` over multiple GPUs (See [here](https://gitlab.mpcdf.mpg.de/ift/nifty/-/merge_requests/993) ). Currently, this already works for MGVI inference, and I will add support for non-linear sampling (geoVI) in the next few days.

If you are interested, feel free to try this development branch. You can find a demo of how to use it [here](https://gitlab.mpcdf.mpg.de/ift/nifty/-/blob/multi_gpu/demos/a_multi_gpu.py?ref_type=heads) (You just need to pass the JAX devices over which you want to map to `optimize_kl`). Please note that I have not yet made any performance tests. Thus if you try it you should have an eye on performance.
